[FEAT] CategoryView 기본 Layout 구현

Suyeon9911 opened the pull request:
## 🟣 관련 이슈

<!-- 관련있는 이슈 번호(#000)을 적어주세요. -->

- close #21

## 🟣 구현/변경 사항 및 이유

<!-- 구현/변경한 내용과 그 이유를 적어주세요. -->
- [x] CategoryCollectionView 레이아웃
- [x]  CategoryCell 구현 
- [x] 카테고리 추가 버튼 & 전체 카테고리 라벨 추가 
- [x] CategoryCoordinator 추가 
 
## 🟣 PR Point

<!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
PR Point에 주석 달아놨습니다 !! 

## 🟣 참고 사항

<!-- 참고할 사항(+스크린샷)이 있다면 적어주세요. -->

셀 안에 더미데이터 모델이랑 이미지는 아직 적용안된 상태입니다! 
에셋 넣으면 수정하도록 하겠습니다! 
왼쪽이 지금까지 레이아웃 잡은거고, 오른쪽은 제가 연습레포에서 더미데이터 넣고 , 이미지도 임시로 넣어본 화면입니당 ! 

<img src ="https://user-images.githubusercontent.com/81313960/149334389-e084db46-0102-4eeb-b6e9-00f1a096b112.png" width = 300 > <img src= "https://user-images.githubusercontent.com/81313960/149334550-5701951d-f45b-4a69-b63f-6fdacb9a40aa.png" width = 300 >

Suyeon9911 committed:
[FEAT] Category CollectionView 레이아웃 (#21)

Suyeon9911 committed:
[ADD] CategoryCoordinator 추가 (#21)

Suyeon9911 committed:
[FEAT] Category Cell 구현 (#21)

Suyeon9911 committed:
Merge remote-tracking branch 'origin/develop' into feature/21-categoryview

Suyeon9911 committed:
[FEAT] 카테고리 추가 버튼 & 전체 카테고리 라벨 추가 (#21)

Suyeon9911 committed:
[ADD] PR Point 주석추가 (#21)

Suyeon9911 committed:
[ADD] Font & Color Extension 적용 (#21)

Suyeon9911 committed:
[FEAT] CategoryView NavigationBar 레이아웃 (#21)

Suyeon9911 committed:
[MERGE] develop branch 병합

Suyeon9911 committed:
[Correct] SwiftLint 룰에 맞게 코드 수정

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+
+class CategoryCollectionViewCell: UICollectionViewCell {
+
+    // MARK: - Vars & Lets Part
```
p5;
나중에 MARK 형식은 어떻게 할건지는 다같이 정리하시죱
해당 주석은 남겨두신 이유가 있나요?

Suyeon9911 replied:
@sujinnaljin  넵 ! 알겠습니다 고냥 전에 쓰던거 쇽샥 마크... 헿

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+
+    private let categoryImageView: UIImageView = {
+        let image = UIImageView()
+        image.image = UIImage(named: "category_icon")
```
p5;
나중에 요 부분은 ImageLiteral 쪽에 빼지는걸 사용하면 될 것 같습니다

Suyeon9911 replied:
@sujinnaljin 넵 !!

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+    // MARK: - Life Cycle Part
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        setLayouts()
```
p5;
이 함수는 extension 쪽에 정의해두고 setBackground 함수는 내부에 정의해두신 이유가 있나요?

Suyeon9911 replied:
@sujinnaljin setLayout 함수를 맨 아래 익스텐션으로 빼야 레이아웃에 대한 코드를 보기 편해서 뺏는데 setBackground 함수는 따로 추가하다보니까 생각못했네용 ! extension 쪽으로 빼는게 좋을까요??

sujinnaljin replied:
엥 왜 제 댓글 없어졌져
뭐가 됐든 상관은 없겠지만, class 선언 내부에 모으거나 extension 에 모으거나 한곳에 정의해주는게 좋을 것 같습니다
제 개취는 레이아웃쪽은 class 내부에 선언할 거 같아요

Suyeon9911 replied:
@sujinnaljin  넵 ! 윤아대장님도 class 내부에 선언하는게 낫다고 하시네용 ! 반영하겠습니다 ~~

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+        contentView.layer.cornerRadius = 6
+    }
+
+    // MARK: - Custom Method Part
```
p3;
필요없는 주석은 정리해주세요

Suyeon9911 replied:
@sujinnaljin  넵 !!

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+
+    // MARK: - Custom Method Part
+    private func setBackground() {
+        contentView.layer.backgroundColor = UIColor.purpleCategory.cgColor
```
p2;
아래같은 코드는 안될까여?
```suggestion
self.backgroundColor = .purpleCategory
```

Suyeon9911 replied:
@sujinnaljin 이 부분 잘 사용하는 법을 잘 몰랐는데 좋은 것 같아요 ! 반영하겠습니다 ~~

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+    private func setBackground() {
+        contentView.layer.backgroundColor = UIColor.purpleCategory.cgColor
+        layer.cornerRadius = 6
+        contentView.layer.cornerRadius = 6
```
p3;
cornerRadius 설정은 따로 두번 따로 해줘야하나요?

sujinnaljin replied:
p4;
cornerRadius 를 background 라고 생각하기 어려운 것 같습니다. 이를 포괄 할 수 있는 함수 네이밍을 해주시거나, 해당 기능을 따로 분리해주시면 좋을 것 같습니다.

Suyeon9911 replied:
@sujinnaljin 제가 찾아봤던 레퍼는 두번 설정해주는데 하나만 해도 되는지 한번 테스트해보고 수정하도록 하겠습니다 !  + 함수 네이밍도 다시 생각해볼게요 !

YoonAh-dev replied:
p4;
layer.cornerRadius만 잡아줘도 Cell 자체 Radius가 제대로 잡힐거예요!

Suyeon9911 replied:
@YoonAh-dev  넵 ! 반영하겠습니다 ~

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    private func setViewHierarchies() {
+        // 📌 이거 여러개 추가하는 extension 어케 쓰죵
```
```swift
 self.contentView.adds([categoryImageView, categoryTitleLabel])
```
처럼 사용하시면 됩니다.

p2;
@YoonAh-dev 님 adds 함수 이름을 `addSubViews` 정도로 변경이 필요할 것 같습니다. 
adds 는 그냥 3인칭 동사같고 subview를 add 한다는 것도 의미 전달이 어려운 것 같습니다.

YoonAh-dev replied:
알겠습니다. 이번에 Cell Base부분에다가 해당 부분 추가해서 머지하겠습니다.

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+
+    private func setViewHierarchies() {
+        // 📌 이거 여러개 추가하는 extension 어케 쓰죵
+        contentView.addSubview(categoryImageView)
```
p4;
몰라서 여쭤보는데 collectionViewCell 에 subview를 추가할때는 self.addSubView(categoryImageView) 하면 안되나요? 꼭 contentView를 접근해야하는건가요??

Suyeon9911 replied:
@sujinnaljin 저도 잘 몰라서 방금 윤아언니한테 물어봤는데 셀 자체에 올릴때는 self.addSubview 가 맞다고 합니다! 수정하겠습니다 !

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+    private func setConstraints() {
+        categoryImageView.snp.makeConstraints {
+            // ✨ 이미지뷰 레이아웃 뭐 잡아줘야하는지 !
+            $0.top.leading.bottom.equalToSuperview().inset(7)
```
p2;
오토레이아웃은 위치와 크기에 대해 값을 결정해줍니다.
top, leading, bottom 만 잡게 되면 width 에 대한 정보가 없게 됩니다. 
따라서 width 를 지정하거나 traliling 을 지정해줘야할 것 같습니다
밑에 것들도 이 개념을 참고해서 다시 작성해주셔야할 것 같습니다 (콘솔에 에러 안떴나요,,??? 제가 잘못알고 있는건쥐,,??)
cc: @YoonAh-dev  님 맞나요?

YoonAh-dev replied:
@sujinnaljin @Suyeon9911 

현재 코드만을 봤을 때, width가 잡히지 않은 상태입니다. width를 정할 수 있도록 trailing를 잡거나, width값을 지정해줘야 합니다!

top, bottom의 경우에도 셀에 height가 잡혀있지 않으면 imageView의 height가 ambiguous하다고 오류가 납니다! ImageView, View의 경우에는 default height, width가 없으니 항상 주의해주세요!

sujinnaljin replied:
윤아님이 말씀주신 
> ImageView, View의 경우에는 default height, width가 없으니 항상 주의해주세요!

는 intrinsicContentSize 와 관련이 있을 듯 합니다.
더 자세한 내용은 아래 링크들을 참고해주세요
[intrinsicContentSize 와 invalidateIntrinsicContentSize()](https://github.com/sujinnaljin/TIL/blob/master/Swift/intrinsicContentSize%20%EC%99%80%20invalidateIntrinsicContentSize().md)
[https://magi82.github.io/ios-intrinsicContentSize/](https://magi82.github.io/ios-intrinsicContentSize/)
https://developer.apple.com/documentation/uikit/uiview/1622600-intrinsiccontentsize

YoonAh-dev replied:
덧붙이자면, subview의 height, width가 안잡혔는데 뷰 위에 얹혀있으면 오류는 안잡히지만 그 안에 들어있는 UI 속성들이 tap이 되지 않는다는 문제가 생기기때문에(보이긴하지만 허상인 느낌..) 항상 사용하실 때 유념해주세요~

Suyeon9911 replied:
@sujinnaljin @YoonAh-dev  넵 !! 감사합니다 !~~ 참고하여 반영하겠씁니다 ㅎㅎ

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryViewController.swift:
```diff
@@ -7,23 +7,184 @@
+
+    // MARK: - UI Component Part
+
+    private lazy var categoryCollectionView: UICollectionView = {
```
p4; 
lazy 로 선언하신 이유가 뭘까요?

Suyeon9911 replied:
@sujinnaljin 저도.. 그 이유를 까먹었네요..? 다시 찾아보도록하겠습니다..

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryViewController.swift:
```diff
@@ -7,23 +7,184 @@
     override func viewDidLoad() {
         super.viewDidLoad()
+        setDelegation()
+        setLayouts()
```
p1;
BaseViewController 에 아래와 같은 함수들이 정의되어있습니다. BaseViewController를 상속받고 해당 함수를 override 해서 사용해주세요
```swift
func render() {
        // Override Layout
    }
    
    func configUI() {
        // View Configuration
    }
```

Suyeon9911 replied:
@sujinnaljin  넵 반영하겠습니다 !

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryViewController.swift:
```diff
@@ -7,23 +7,184 @@
+    private func setNavigationBar() {
+        title = "전체 카테고리"
+        // 여기 폰트적용 어떻게 하나용
+        navigationController?.navigationBar.titleTextAttributes = [.font: UIFont(name: "Pretendard-Bold", size: 16)]
```
p3;
요렇게도 사용할 수 있습니다
```suggestion
navigationController?.navigationBar.titleTextAttributes = [.font: UIFont.font(.pretendardBold, ofSize: 16)]
```
그나저나 저 코드 안먹히나요?

Suyeon9911 replied:
@sujinnaljin 안먹히지는 않는데 경고가 떠서 ..!! 여쭤봤습니다 .. 제안해주신 코드 반영했더니 사라졌어요 ~~  최고최고

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryViewController.swift:
```diff
@@ -7,23 +7,184 @@
+
+extension CategoryViewController: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        // 📌 이부분 셀 width 비율에 맞춰서 해주고 싶은데 이 방법밖에 생각이 안나서 더 좋은 방법이 있다면 알려주세요!
```
375가 cellWidth 라는건가요?
```swift
class MyCell {
   enum Size {
   static let width: CGFloat = 375 // 물론 계산해야합니다
  }
}

let widthRatio: CGFloat = 345/MyCell.Size.width
```
처럼 사용하는건 어떨까요?

Suyeon9911 replied:
@sujinnaljin collectionView의 width가 375이고, cell의 width 가 345인데 기기마다 셀의 크기가 다르게 보여서 어떻게 해줄까 하다가.. collectionview에 대한 cell의 비율로 계산해서 넣었던 거였습니다 1!!

sujinnaljin replied:
cell의 width 가 345인데 기기마다 셀의 크기가 다르게 보여서,,,???
cell 의 width 가 고정이라면
```swift
CGSize(width: 345, height: 56)
```
처럼 사용하면 되는것 아닌가요,,,,,,,,,,,??
지금 수연님이 사용하신 코드는 디바이스에 따라 가변적이게 되는거 아닌가요,,???!! ㅇㅂㅇ

Suyeon9911 replied:
@sujinnaljin  크기를 고정으로 해주니 기기마다 셀의 크기가 달라져서 .. 어떻게 하믄 좋을까요..? @YoonAh-dev  대장님..

YoonAh-dev replied:
셀이 width마다 바뀌는 것이라면 전체 Screen Width에서 양 옆 마진을 빼면 되지 않을까요?
```swift
UIScreen.main.bounds.size.width - 마진값 * 2
```

Suyeon9911 replied:
@YoonAh-dev  오... 일단 이거대로 적용하니까 원하는대로 되었습니다...! !!! !

YoonAh-dev commented on Havit/Havit/Coordinator/CategoryCoordinator.swift:
```diff
@@ -0,0 +1,25 @@
+
+import Foundation
+
+final class CategoryCoordinator: BaseCoordinator {
```
오 잘 적용하셨네요?👍

Suyeon9911 replied:
@YoonAh-dev  감사합니다.. 대장님의 칭찬.. 냠냠

YoonAh-dev commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+
+    private let categoryTitleLabel: UILabel = {
+        let label = UILabel()
+        label.font = .font(FontName.pretendardMedium, ofSize: 14)
```
```suggestion
        label.font = .font(.pretendardMedium, ofSize: 14)
```
이렇게 쓰셔도 무방합니다.!

Suyeon9911 replied:
@YoonAh-dev 오호! 새로 알아갑니다 적용 갈길게요

YoonAh-dev commented on Havit/Havit/Screens/Category/View/CategoryViewController.swift:
```diff
@@ -7,23 +7,184 @@
+        default:
+            break
+        }
+    }
```
p2;

bind()함수를 만들어서(저번에 texture-MVVM에서 했던 것처럼) button tapped코드를 RxCocoa형식으로 바꿀 수 있을 거 같네요! tap하는 행위에 의해서 ViewModel로 전달된 부분은 없을 거 같아서 `button.rx.tap`정도 구현 가능할 거 같습니다.

Suyeon9911 replied:
@YoonAh-dev  넵 ! 반영하였습니다  1086e4c

Suyeon9911 committed:
[Chore] 폰트 적용 코드 수정

Suyeon9911 committed:
[CHORE] 주석 코드 컨벤션에 맞게 수정

Suyeon9911 committed:
[CORRECT] Lint 룰에 맞게 수정

Suyeon9911 committed:
[CHORE] Cell width 적용 코드 수정

Suyeon9911 committed:
[CHORE] 코드리뷰 반영하여 레이아웃 수정

Suyeon9911 committed:
[CORRECT] Lint 룰에 맞게 수정

Suyeon9911 committed:
[CHORE] Font 적용 코드 수정

Suyeon9911 committed:
Merge branch 'develop' of https://github.com/TeamHavit/Havit-iOS into feature/21-categoryview

Suyeon9911 committed:
[CHORE] adds 함수 네이밍 변경

beansbin reviewed (commented):
Coordinator 적용 방법 잘 보고 갑니다🙇🏻‍♂️

Suyeon9911 committed:
[CHORE] 코드리뷰 반영

- CategoryCollectionViewController.swift : BaseViewController 상속
- CategoryCollectionViewCell.swift: BaseCollectionViewCell 상속

sujinnaljin commented on Havit/Havit/Screens/Category/View/CategoryCollectionViewCell.swift:
```diff
@@ -0,0 +1,77 @@
+        setBackground()
+    }
+
+    required init?(coder: NSCoder) {
```
p4;
혹시라도 해당 생성자가 호출되지 않도록 아예 아래와 같이 접근을 막아버릴 수도 있습니다~ (호출되면 무조건 fatal Error 가 나기 때문)
```swift
    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
```

Suyeon9911 replied:
@sujinnaljin  오호 !! 몰랐는데 .. 하나 더 배워갑니다 ..

sujinnaljin replied:
코드리뷰가 백개가 되길 원한다고 들었는데,,! 👀
반영한 사항에 대해서는 77f704b 처럼 커밋 번호로 남겨주시면 리뷰가 백개도 되고 리뷰어가 실제로 반영되었는지 확인도 가능해서 좋을 것 같습니다

Suyeon9911 replied:
@sujinnaljin 헐 대박 꿀팁 ... 넘 좋습니다 ... 앞으로 커밋 번호 갈기도록 하겠습니다 !

Suyeon9911 committed:
[CHORE] 코드리뷰 반영

- fatal error

YoonAh-dev commented on Havit/Havit/Screens/Category/View/CategoryViewController.swift:
```diff
@@ -7,22 +7,170 @@
+        view.addSubViews([categoryCollectionView, categoryCountLabel, addButton])
+
+        categoryCountLabel.snp.makeConstraints {
+            $0.top.equalTo(self.view.safeAreaLayoutGuide.snp.top).inset(22)
```
p3;
```suggestion
            $0.top.equalTo(view.safeAreaLayoutGuide.snp.top).inset(22)
```
현재 ViewController위에 올라가있는 view 이외에 다른 view가 없기 때문에 self없이 view.safeAreaLayoutGuide로 써줘도 될 거 같습니다.

Suyeon9911 replied:
@YoonAh-dev 넵 ! 반영했습니다 ! ! --> 커밋번호를 따라가세요 슉슉슉슉.. 25fd6f5

Suyeon9911 committed:
[CHORE] RxCocoa 적용

- CategoryViewController: buttonTapped 함수를 RxCocoa형식으로 수정

Suyeon9911 committed:
[CHORE] 코드리뷰 반영

- CategoryViewController.swift : safearea 레이아웃 코드 수정

Suyeon9911 committed:
[FEAT] Main 화면전환 구현

YoonAh-dev commented on Havit/Havit/Screens/Category/View/CategoryViewController.swift:
```diff
@@ -7,22 +7,176 @@
+                self?.coordinator?.performTransition(to: .main)
+            })
+            .disposed(by: disposeBag)
+
```
p1;
@Suyeon9911 빈칸이 있네요.🥲

Suyeon9911 replied:
@YoonAh-dev  반 . 영 . 했 . 습. 니. 다 . ----> 721dcbb

Suyeon9911 committed:
[CHORE] 코드리뷰 반영

YoonAh-dev reviewed (approved):
늦은 시간까지 고생하셨습니다.
코드가 많이 깔끔해졌네요 👍

noah0316 reviewed (approved):
고생하셨습니다~~!! 선배 수고꽈리!!

sujinnaljin reviewed (approved):
고생 많으셨습니다! 사실 처음 pr 도 잘 날려주시긴 했지만, 훨씬 잘 읽히네여!!
